test: remove wallet dependency from feature_nulldummy.py

ayush933 opened the pull request:
This PR enables one of the non-wallet functional tests (`feature_nulldummy.py`) to be run even with the Bitcoin Core wallet disabled.

Commit 1: removes wallet dependency and `test_runner.py` is edited to make sure the test only runs once.
Commit 2: the functions `create_transaction()` and `create_raw_transaction()` in `blocktools.py` are no longer needed and hence removed.

MarcoFalke commented on test/functional/feature_nulldummy.py:
```diff
@@ -82,31 +82,59 @@ def run_test(self):
+        output = {self.ms_address: 49}
+        rawtx = self.nodes[0].createrawtransaction([input], output)
+        signedtx = self.nodes[0].signrawtransactionwithkey(hexstring=rawtx, privkeys=[self.nodes[0].PRIV_KEYS[0][1]])
+        test1txs = [tx_from_hex(signedtx["hex"])]
```
I am wondering if it helps if there was a local `create_transaction` helper? As this is creating a simple chain of txs, the vout could be hardcoded to `0` in the helper, also the node could be hardcoded to `self.nodes[0]`, as there is only one node.

So you'd call `self.create_transaction(coinbase_txid[0], self.ms_address, amount=49, privkey=self.nodes[0].PRIV_KEYS[0][1])` here.

The benefit would be less code and less symbols in this scope.

Thoughts?

maflcko reviewed (approved):
Looks good. Left an idea

laanwj commented:
Concept ACK, thanks for working on disentangling more tests from the wallet.

MarcoFalke commented on test/functional/feature_nulldummy.py:
```diff
@@ -55,58 +60,87 @@ def set_test_params(self):
-            coinbase_txid.append(self.nodes[0].getblock(i)['tx'][0])
+            block = self.nodes[0].getblock(i)
+            coinbase_txid.append(block['tx'][0])
+            block_hash.append(block["hash"])
```
Is this needed? In the previous commit you didn't have to pass in the scriptPubKey for coinbase spends with the deterministic keys.

MarcoFalke replied:
Oh I see. This is due to passing the input even though it is not needed. What do you think about passing `None` in this case?

See this draft diff:


```diff
diff --git a/test/functional/feature_nulldummy.py b/test/functional/feature_nulldummy.py
index e860a7fc36..90e0887b08 100755
--- a/test/functional/feature_nulldummy.py
+++ b/test/functional/feature_nulldummy.py
@@ -60,19 +60,13 @@ class NULLDUMMYTest(BitcoinTestFramework):
             '-par=1',  # Use only one script thread to get the exact reject reason for testing
         ]]
 
-    def create_transaction(self, *, txid, inAmount=None, spk=None, rs=None, ws=None, addr, amount, privkey):
+    def create_transaction(self, *, txid, prev_utxo_details=None, addr, amount, privkey):
         input = {"txid": txid, "vout": 0}
-        if inAmount is not None:
-            input["amount"] = inAmount
-        if spk is not None:
-            input["scriptPubKey"] = spk
-        if rs is not None:
-            input["redeemScript"] = rs
-        if ws is not None:
-            input["witnessScript"] = ws
         output = {addr: amount}
         rawtx = self.nodes[0].createrawtransaction([input], output)
-        signedtx = self.nodes[0].signrawtransactionwithkey(rawtx, [privkey], [input])
+        # Details only needed for scripthash or witness spends
+        input = None if not prev_utxo_details else [{**input,**prev_utxo_details}]
+        signedtx = self.nodes[0].signrawtransactionwithkey(rawtx, [privkey], input)
         return tx_from_hex(signedtx["hex"])
 
     def run_test(self):
@@ -101,24 +95,22 @@ class NULLDUMMYTest(BitcoinTestFramework):
 
         self.log.info(f"Test 1: NULLDUMMY compliant base transactions should be accepted to mempool and mined before activation [{COINBASE_MATURITY + 3}]")
         test1txs = [self.create_transaction(txid=coinbase_txid[0], addr=self.ms_address, amount=49,
-                                            spk=self.nodes[0].getrawtransaction(coinbase_txid[0], True, block_hash[0])["vout"][0]["scriptPubKey"]["hex"],
                                             privkey=self.nodes[0].PRIV_KEYS[0][1])]
         txid1 = self.nodes[0].sendrawtransaction(test1txs[0].serialize_with_witness().hex(), 0)
         spk = test1txs[0].vout[0].scriptPubKey.hex()
-        test1txs.append(self.create_transaction(txid=txid1, spk=spk,
-                                                rs=cms["redeemScript"],
+        test1txs.append(self.create_transaction(txid=txid1, prev_utxo_details={"scriptPubKey":spk,
+                                                "redeemScript":cms["redeemScript"]},
                                                 addr=self.ms_address, amount=48,
                                                 privkey=self.privkey))
         txid2 = self.nodes[0].sendrawtransaction(test1txs[1].serialize_with_witness().hex(), 0)
         test1txs.append(self.create_transaction(txid=coinbase_txid[1],
-                                                spk=self.nodes[0].getrawtransaction(coinbase_txid[1], True, block_hash[1])["vout"][0]["scriptPubKey"]["hex"],
                                                 addr=self.wit_ms_address, amount=49,
                                                 privkey=self.nodes[0].PRIV_KEYS[0][1]))
         txid3 = self.nodes[0].sendrawtransaction(test1txs[2].serialize_with_witness().hex(), 0)
         self.block_submit(self.nodes[0], test1txs, accept=True)
 
         self.log.info("Test 2: Non-NULLDUMMY base multisig transaction should not be accepted to mempool before activation")
-        test2tx = self.create_transaction(txid=txid2, spk=spk, rs=cms["redeemScript"],
+        test2tx = self.create_transaction(txid=txid2, prev_utxo_details={"scriptPubKey":spk, "redeemScript":cms["redeemScript"]},
                                           addr=self.ms_address, amount=47,
                                           privkey=self.privkey)
         invalidate_nulldummy_tx(test2tx)
@@ -128,7 +120,7 @@ class NULLDUMMYTest(BitcoinTestFramework):
         self.block_submit(self.nodes[0], [test2tx], accept=True)
 
         self.log.info("Test 4: Non-NULLDUMMY base multisig transaction is invalid after activation")
-        test4tx = self.create_transaction(txid=test2tx.hash, spk=spk, rs=cms["redeemScript"],
+        test4tx = self.create_transaction(txid=test2tx.hash, prev_utxo_details={"scriptPubKey":spk, "redeemScript":cms["redeemScript"]},
                                           addr=self.address, amount=46,
                                           privkey=self.privkey)
         test6txs = [CTransaction(test4tx)]
@@ -137,8 +129,8 @@ class NULLDUMMYTest(BitcoinTestFramework):
         self.block_submit(self.nodes[0], [test4tx], accept=False)
 
         self.log.info("Test 5: Non-NULLDUMMY P2WSH multisig transaction invalid after activation")
-        test5tx = self.create_transaction(txid=txid3, spk=test1txs[2].vout[0].scriptPubKey.hex(),
-                                          inAmount=49, ws=wms["redeemScript"],
+        test5tx = self.create_transaction(txid=txid3, prev_utxo_details={"scriptPubKey":test1txs[2].vout[0].scriptPubKey.hex(),
+                                          "amount":49, "witnessScript":wms["redeemScript"]},
                                           addr=self.wit_address, amount=48,
                                           privkey=self.privkey)
         test6txs.append(CTransaction(test5tx))

kouloumos commented on test/functional/feature_nulldummy.py:
```diff
@@ -82,31 +91,45 @@ def run_test(self):
         txid1 = self.nodes[0].sendrawtransaction(test1txs[0].serialize_with_witness().hex(), 0)
-        test1txs.append(create_transaction(self.nodes[0], txid1, self.ms_address, amount=48))
+        spk = test1txs[0].vout[0].scriptPubKey.hex()
+        test1txs.append(self.create_transaction(txid=txid1, prev_utxo_details={"scriptPubKey": spk,
+                                                "redeemScript": cms["redeemScript"]},
```
nit: As the same `prev_utxo_details` are used more than once, I think it would be beneficial to the test's readability if you assigned them at the beginning of the test with something like 
`ms_address_unlock_details = {"scriptPubKey": spk,
                                                "redeemScript": cms["redeemScript"]}`

kouloumos commented on test/functional/feature_nulldummy.py:
```diff
@@ -55,22 +60,26 @@ def set_test_params(self):
+        cms = self.nodes[0].createmultisig(1, [self.pubkey])
+        wms = self.nodes[0].createmultisig(1, [self.pubkey], 'p2sh-segwit')
+        self.ms_address = cms["address"]
+        self.wit_address = getnewdestination(address_type='p2sh-segwit')[2]
```
nit, maybe move address to where it is first used

kouloumos reviewed (commented):
ACK 9a783a9fd76ee5350d64ed1ed5b658c530913653 with some minor comments that i believe could increase readability.

I was also thinking that using the MiniWallet could potentially simplify the test and I tried to validate my intuition, but it seems that the changes needed do not justify the idea.

ayush933 committed:
test: remove wallet dependency from feature_nulldummy.py

This test can now be run even with the Bitcoin Core wallet disabled.

ayush933 committed:
remove unused functions

the functions `create_transaction()` and `create_raw_transaction()` were no longer used hence removed.

ayush933 commented:
Thanks a lot @MarcoFalke and @kouloumos for the thorough reviews.
Addressed the reviews and made the necessary changes.
